### PR TITLE
Changes to MessageEvent

### PR DESCRIPTION
**Pull request by windy1**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/1118) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/548) | [Vanilla](https://github.com/SpongePowered/SpongeVanilla/pull/229) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/559)

This PR makes some substantial changes to how MessageEvents are handled. Instead of passing around a `Text` object, the event now contains a `MessageFormatter` which can be used to modify the event's message in a more reliable and modular fashion. This will make interactions and compatibility between chat plugins a lot easier. Previous methods like `getOriginalMessage`, `getRawMessage`, `getMessage`, and `setMessage` still behave as expected except instead of Optionals, a non-null Text is always returned. Developers should use `setMessageCancelled` to cancel the sending instead of setting the message to null.

----

Some example uses:

Let's say I would like to insert a "group" prefix to the message but I also would like to keep existing prefixes that other plugins may or may not have already applied. I can do something like:

```java
class GroupApplier extends SimpleTextTemplateApplier {
    public GroupApplier() {
        super(of('[', arg("group"), "] "));
    }
}
```

```java
@Listener
void onChat(MessageChannelEvent.Chat event) {
    GroupApplier applier = new GroupApplier();
    applier.setParameter("group", Text.builder("admin").color(TextColors.RED));
    event.getFormatter().getHeader().insert(0, applier);
}
```

On an unaltered message, this will produce something equivalent to:

`[<red>admin</red>] <NAME> MESSAGE`

Now lets say another plugin wants to override that prefix and make it say "mod" instead of "admin". Well then they can do:

```java
event.getHeader().forEach(GroupApplier.class, applier -> applier.setParameter("group", Text.builder("mod").color(TextColors.AQUA));
```

These are only a few of the many uses this PR will provide.

**Note:** This PR depends on https://github.com/SpongePowered/SpongeAPI/pull/1079. https://github.com/SpongePowered/SpongeAPI/pull/1079 should be merged into master *and then into bleeding* before this PR is merged.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java [hunk 1]
```diff
@@ -24,37 +24,262 @@
...
 /**
  * Describes events when a involving a {@link Text} message.
  */
+@ImplementedBy(AbstractMessageEvent.class)
 public interface MessageEvent extends Event {
 
     /**
+     * Parameter for header section in header partition.
+     */
+    String PARAM_MESSAGE_HEADER = "header";
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: https://github.com/SpongePowered/SpongeAPI/pull/1114#discussion_r54342803

File: src/main/java/org/spongepowered/api/event/message/MessageEvent.java [hunk 2]
```diff
@@ -24,37 +24,262 @@
...
+
+    /**
+     * Parameter for body section in body partition.
+     */
+    String PARAM_MESSAGE_BODY = "body";
+
+    /**
+     * Parameter for footer section in footer partition.
+     */
+    String PARAM_MESSAGE_FOOTER = "footer";
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/event/message/MessageEvent.java]: https://github.com/SpongePowered/SpongeAPI/pull/1114#discussion_r54342802

File: src/main/java/org/spongepowered/api/text/transform/TextFormatter.java [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+     */
+    boolean add(E element);
+
+    /**
+     * Adds the specified elements to the end of this formatter.
+     *
+     * @param elements Elements to add
+     * @return True if the formatter changed as a result of the call
+     */
+    boolean add(Collection<E> elements);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/transform/TextFormatter.java]: https://github.com/SpongePowered/SpongeAPI/pull/1114#discussion_r54342840

**Comment by ST-DDT** [comment]
@windy1 Instead of continuously closing and opening new PRs you could just update them. 
This makes tracking the discussion a lot easier.

**Comment by windy1** [comment]
@ST-DDT Trust me it's not intentional :P I keep messing up rebasing the feature/text-template branch....

**Comment by JBYoshi** [comment]
@windy1 You can undo the changes with `git reflog` and then force-push the old versions.

**Comment by ST-DDT** [comment]
Well you can still fix / recreate your changes and then force push to the branch associated with the PRs. Even if the original branch is horribly broken a simple force push with a working state will also fix the PR.

**Comment by windy1** [comment]
Okay -- I'd like to get this merged ASAP now that TextTemplates are in. I'm still partial to "header, body, footer" but if someone can make a compelling argument against it other than "I like 'x' better" I'll change it.

**Pull request merged**